### PR TITLE
Upgrade antd to version 4

### LINKED ISSUE #4678: Update antd to v4.x
See migration guide at https://ant.design/docs/react/migration-v4

Some relevant improvements I spotted:
  - Better Form handling, no more getFieldDecorator etc. (https://ant.design/components/form/v3/#Remove-Form.create)
  - Icons are separate components, allowing tree shaking and a smaller bundle (https://ant.design/docs/react/migration-v4#Icon-upgrade)
  - Tree, Select, TreeSelect, AutoComplete use virtual scrolling

We could then also use the `loading` prop of Selects instead of the `notFoundContent` workaround introduced in https://github.com/scalableminds/webknossos/pull/4671

### LINKED ISSUE #5165: Text gets selected when adjusting histogram sliders
## Context
In Firefox, when viewing a dataset and editing the displayed intensity range using the sliders below the histogram, text can get selected when the mouse wanders down during dragging.

## Expected Behavior
The text should not get selected when dragging a UI element, even when the mouse wanders down during dragging.

## Current Behavior
Text gets selected.

image

## Your Environment for bug
Firefox 85, Windows 10

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
Upgrade UI library antd to version 4, creating a slightly more modern look and behavior of many UI elements and reducing technical debt.

### URL of deployed dev instance (used for testing):
- https://upgradeantd.webknossos.xyz

### Steps to test:
- see comment below, https://github.com/scalableminds/webknossos/pull/5350#issuecomment-814065777

### Issues:
- fixes #4678
- fixes #5165 🎉 

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [X] Ready for review

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
List of components to be tested in dev deployment

- [ ] `accept_invite_view`
- [ ] `Auth token view`
- [ ] `Password Change View`
- [ ] `Finish Reset Password view`
- [ ] Login View
- [ ] Registration Form
- [ ] Spotlight Registration Form (Registration Form used within the navbar afaik)
- [ ] Start Reset Password View
- [ ] Dataset Add Boss View
- [ ] Add Foreign Dataset View
- [ ] Add Neuroglancer Dataset View
- [ ] Dataset Add View
- [ ] Dataset Upload View
- [ ] Job List View
- [ ] onboarding
- [ ] Organization edit view
- [ ] project create view
- [ ] project list view
- [ ] Script Create View
- [ ] Script List view
- [ ] Project Progess Report View
- [ ] Team Selection Form
- [ ] Task Annotation View
- [ ] Task create bulk view
- [ ] Task Create Form View & Task Create View
- [ ] Task List View
- [ ] Task Search Form
- [ ] Recommended Configuration View
- [ ] Admin Task Type Create View

**Comment by grittaweisheit** [comment]
[ ] version_view
[ ] user_settings_view
[ ] setting_input_view
[ ] dataset_settings_view
[ ] trees_tab_view
[ ] tree_hierarchy_view
[ ] meshes_view
[ ] mapping_info_view
[ ] dataset_info_tab_view
[ ] comment_tab_view
[ ] tracing_layout_view
[ ] volume_actions_view
[ ] view_dataset_actions_view
[ ] user_scripts_modal_view
[ ] tracing_actions_view
[ ] share_view_dataset_modal_view
[ ] share_modal_view
[ ] merger_modal_view
[ ] dataset_position_view
[ ] navbar
[ ] shared_annotations_view
[ ] publication_details_view
[ ] explorative_annotations_view
[ ] dataset_view
[ ] dataset_import_view
[ ] dashboard_view
[ ]

**Comment by fm3** [comment]
Great stuff! I noticed a few small problems in my assigned views:
- [ ] task list view, clicking the plus/minus icon in the first column again does not collapse immediately, only after third click
- [ ] create task view, dataset input is autofocused, way down in view
- [ ] version view: Tabs need padding
- [ ] dataset settings view: icon placement is off (find data, unlink fallback layer, reload) AsyncIconButton
- [ ] dataset settings view: icons don’t have tooltip (find data, unlink fallback layer, reload) AsyncIconButton
- [ ] trees_tab_view: tree visibility checkbox placement is off (too low)
- [ ] trees_tab_view: hover highlight partly occludes the checkbox of the tree above
- [ ] tracing_actions_view: "autoSaveLayouts ? " shown in UI, along both buttons (should be if-else I suppose?) Probably missing curly braces?
- [ ] share_view_dataset_modal_view: should not say "share this Tracing". Should say "Dataset" (same on master, though)
- [ ] merger_modal_view: Lines should have padding

File: frontend/javascripts/dashboard/dataset/dataset_import_view.js [hunk 1]
```diff
@@ -223,7 +225,11 @@ class DatasetImportView extends React.PureComponent<Props, State> {
         datasetClone.dataSource.status = dataset.dataSource.status;
         dataset = (datasetClone: APIDataset);
       }
-      this.props.form.setFieldsValue({
+      const form = this.formRef.current;
+      if (!form) {
+        return;
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply frontend/javascripts/dashboard/dataset/dataset_import_view.js:230]: This is the "template" @MichaelBuessemeyer introduced everywhere `form` is accessed in a React class component. I agree that it would be better to throw an error here, instead of failing silently :+1:

File: frontend/javascripts/oxalis/view/td_view_controls.js [hunk 2]
```diff
@@ -23,7 +22,7 @@ function TDViewControls({ isRefreshingIsosurfaces, volumeTracing }: Props) {
     }
   }
   return (
-    <ButtonGroup id="TDViewControls">
+    <Space size={0} id="TDViewControls">
```
Comment (grittaweisheit) [inline frontend/javascripts/oxalis/view/td_view_controls.js:25]: This does not align and unify the buttons's sizes and the spaces don't look pretty. Maybe we can fix this with a css class. One is already in the _tracing_view.less, but that's not aligning and unifying.

File: frontend/javascripts/admin/task/task_search_form.js [hunk 3]
```diff
@@ -80,7 +81,11 @@ class TaskSearchForm extends React.Component<Props, State> {
         ? this.props.initialFieldValues
         : this.state.fieldValues;
     if (_.size(fieldValues) > 0) {
-      this.props.form.setFieldsValue(fieldValues);
+      const form = this.formRef.current;
+      if (!form) {
+        return;
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/admin/task/task_search_form.js:86]: Maybe log an error here, as this behaviour is unexpected and would be silent otherwis. But I think throwing an error here or showing a toast is too much as this part only sets the initial values. It should not completely block the user.

File: frontend/javascripts/admin/task/task_search_form.js [hunk 4]
```diff
@@ -99,16 +104,13 @@ class TaskSearchForm extends React.Component<Props, State> {
...
-    if (event) {
-      event.preventDefault();
+  handleFormSubmit = (isRandom: boolean, onFinishCallback: QueryObject => Promise<void>) => {
+    const form = this.formRef.current;
+    if (!form) {
+      return;
     }
 
-    this.props.form.validateFields((err, formValues: TaskFormFieldValues) => {
+    form.validateFields().then(formValues => {
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/admin/task/task_search_form.js:113]: Looks too me like this can be done with `onFinish` not `onSubmit`. That way this method would be smaller and easier to understand.
Using `onFinish` ensures that all fields are valid and therefor a `form.validateFields()` check is not needed

File: frontend/javascripts/admin/task/task_search_form.js [hunk 5]
```diff
@@ -142,98 +144,89 @@ class TaskSearchForm extends React.Component<Props, State> {
...
     const { isLoading } = this.props;
-    const { getFieldDecorator } = this.props.form;
     const formItemLayout = {
       labelCol: { span: 5 },
       wrapperCol: { span: 19 },
     };
 
     return (
-      <Form onSubmit={evt => this.handleSearchFormSubmit(false, evt)}>
+      <Form onFinish={() => this.handleSearchFormSubmit(false)} ref={this.formRef}>
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/admin/task/task_search_form.js:173]: Let's rename all `...FormSubmit` methods to `...FormFinish` to indicate, that the `onFinish` method is used and not manual validation is needed.

File: frontend/javascripts/dashboard/dataset/dataset_import_view.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (philippotto) [reply frontend/javascripts/dashboard/dataset/dataset_import_view.js]: Ok, cool. I changed it in this particular line, since this seemed like it could cause a weird state theoretically. For the other places, I left it as is. It's probably not important in the end, anyway :pray:

File: frontend/javascripts/admin/tasktype/task_type_create_view.js [hunk 6]
```diff
@@ -121,7 +123,11 @@ class TaskTypeCreateView extends React.PureComponent<Props, State> {
     };
     delete formValues.settings.resolutionRestrictions;
 
-    this.props.form.setFieldsValue(formValues);
+    const form = this.formRef.current;
+    if (!form) {
+      return;
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/admin/tasktype/task_type_create_view.js:128]: Consider adding-throwing an error here like in the other files.

File: frontend/javascripts/admin/user/experience_modal_view.js [hunk 7]
```diff
@@ -1,6 +1,7 @@
 // @flow
 
-import { Modal, Button, Tooltip, Icon, Table, InputNumber, Tag, Badge } from "antd";
+import { Modal, Button, Tooltip, Table, InputNumber, Tag, Badge } from "antd";
+import { CloseOutlined, DeleteOutlined, RollbackOutlined } from "@ant-design/icons";
 import * as React from "react";
 import _ from "lodash";
```
Comment (grittaweisheit) [inline frontend/javascripts/admin/user/experience_modal_view.js:7]: This Modal does not look well. The experience entry with domain name, value and bin icon is not aligned.

File: frontend/javascripts/admin/tasktype/task_type_create_view.js [hunk 8]
```diff
@@ -226,131 +225,144 @@ class TaskTypeCreateView extends React.PureComponent<Props, State> {
...
-                  </Radio>
-                  <Radio value="hybrid" disabled={isEditingMode}>
-                    Skeleton and Volume
-                  </Radio>
-                </RadioGroup>,
-              )}
+            <FormItem
+              name="tracingType"
+              label="Annotation Type"
+              onChange={() => this.forceUpdate()}
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/admin/tasktype/task_type_create_view.js:236]: Looks to me like this was added. Why is this needed?

File: frontend/javascripts/admin/tasktype/task_type_create_view.js [hunk 9]
```diff
@@ -226,131 +225,144 @@ class TaskTypeCreateView extends React.PureComponent<Props, State> {
...
+            >
+              {({ getFieldValue }) =>
+                getFieldValue(["settings", "resolutionRestrictionsForm", "shouldRestrict"]) ? (
+                  <div style={{ marginLeft: 24 }}>
+                    <FormItem
+                      name={["settings", "resolutionRestrictionsForm", "min"]}
+                      hasFeedback
+                      label="Minimum"
+                      style={{ marginBottom: 6 }}
+                      rules={[{ validator: isValidMagnification }]}
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/admin/tasktype/task_type_create_view.js:346]: Maybe this is where the bug in this view is from / happening. For more, see my comment that lists the broken views.

**Comment by philippotto** [comment]
> * [x]  task list view, clicking the plus/minus icon in the first column again does not collapse immediately, only after third click

I think I fixed this with 4898076bf. @MichaelBuessemeyer This was due to an old workaround you built for the antd tables. I think, the "ignoreUpdate" field wasn't needed anymore (probably fixed by the version update). I couldn't produce any issues with the new code, so everything should be fine. I mainly tested expanding and collapsing multiple rows. If you remember that there were other problematic scenarios, let me know, but it's probably good now.

**Review by grittaweisheit — COMMENTED** [review]
I saw some issues with the former Button Groups that appear when simply using a `Space` with `size=0`. Someone already introduced a css class that reproduces the looks of the old `ButtonGroup` in `_tracing_view.less`. This doesn't fully replace the ButtonGroup, as the childrens's heights aren't unified and they are not always aligned in the right way. Maybe we could improve the css group class or fix the height and align issues directly where the button groups are placed.

**Comment by philippotto** [comment]
> I saw some issues with the former Button Groups that appear when simply using a `Space` with `size=0`. Someone already introduced a css class that reproduces the looks of the old `ButtonGroup` in `_tracing_view.less`. This doesn't fully replace the ButtonGroup, as the childrens's heights aren't unified and they are not always aligned in the right way. Maybe we could improve the css group class or fix the height and align issues directly where the button groups are placed.

Yes, that was me :)

Which browser do you use? With Chrome and Firefox it looks like this on Linux:
image

Seems alright to me. How does it look for you?

**Review by MichaelBuessemeyer — COMMENTED** [review]
It looks like the following components do not work anymore:
- Statistics -> Open Tasks View => The table does not show any results although there are open tasks that can be seen in the project progress view
- [ ] Admin -> Task List View: The Row Expansion Icon is broken :man_shrugging: 
- [x] Admin -> Tasks -> Add Tasks (Bulk Create View): Dropping a file with the wrong extension does not work (also on master). It just fails silently. -> TODO: Create an Issue that shows an error toast, when the file extension is not supported.
- [x] Admin -> Task Type -> Create View: Under The suboptions of Restrict Resolutions load infinite time for validation.

**Comment by daniel-wer** [comment]
[ ] Admin -> Task Type -> Create View: Under The suboptions of Restrict Resolutions load infinite time for validation.

Did you check this with the newest code? I fixed an issue with the validation that was introduced in some places.

**Comment by grittaweisheit** [comment]
> Yes, that was me :)
> 
> Which browser do you use? With Chrome and Firefox it looks like this on Linux:
> image
> 
> Seems alright to me. How does it look for you?

Here it looks very good, but in the td view controls the Button to load isosurfaces is slightly lower than the others. Maybe that's not really bad, but I can't unsee it. (I'm on Linux with Firefox)
With `Space`s:
Screenshot from 2021-04-06 18-18-24

With the group:
Screenshot from 2021-04-06 18-21-40

**Comment by MichaelBuessemeyer** [comment]
>     * [ ]  Admin -> Task Type -> Create View: Under The suboptions of Restrict Resolutions load infinite time for validation.
> 
> 
> Did you check this with the newest code? I fixed an issue with the validation that was introduced in some places.

I just tested local and it works perfectly. :+1: I will add the `hasFeedback` property so that they also get an error/success icon.

**Comment by MichaelBuessemeyer** [comment]
> I will add the `hasFeedback` property so that they also get an error/success icon.

Never mind. I am blind :see_no_evil:

**Comment by MichaelBuessemeyer** [comment]
> > * [x]   task list view, clicking the plus/minus icon in the first column again does not collapse immediately, only after third click
> 
> I think I fixed this with [4898076](https://github.com/scalableminds/webknossos/commit/4898076bfef9d36b47e627219768e2df2bf56212). @MichaelBuessemeyer This was due to an old workaround you built for the antd tables. I think, the "ignoreUpdate" field wasn't needed anymore (probably fixed by the version update). I couldn't produce any issues with the new code, so everything should be fine. I mainly tested expanding and collapsing multiple rows. If you remember that there were other problematic scenarios, let me know, but it's probably good now.

The icon that is used for expanding / collapsing is still broken for me (Firefox, Linux) :confused:

File: frontend/javascripts/admin/task/task_search_form.js (hunk 5 above)
Comment (philippotto) [reply frontend/javascripts/admin/task/task_search_form.js]: Makes sense. Are you doing that?

**Comment by philippotto** [comment]
> > > * [x]    task list view, clicking the plus/minus icon in the first column again does not collapse immediately, only after third click
> > 
> > 
> > I think I fixed this with [4898076](https://github.com/scalableminds/webknossos/commit/4898076bfef9d36b47e627219768e2df2bf56212). @MichaelBuessemeyer This was due to an old workaround you built for the antd tables. I think, the "ignoreUpdate" field wasn't needed anymore (probably fixed by the version update). I couldn't produce any issues with the new code, so everything should be fine. I mainly tested expanding and collapsing multiple rows. If you remember that there were other problematic scenarios, let me know, but it's probably good now.
> 
> The icon that is used for expanding / collapsing is still broken for me (Firefox, Linux)

Hmm, did you test the newest version? The dev instance is probably not up to date. If there are still issues, let's have a call to talk about this.

**Comment by fm3** [comment]
> > The icon that is used for expanding / collapsing is still broken for me (Firefox, Linux)
> 
> Hmm, did you test the newest version? The dev instance is probably not up to date. If there are still issues, let's have a call to talk about this.

What exactly do you mean by broken? The behavior is fine for me now. However, the icon itself looks weird. There is a fancy animation combining the bars of the plus sign to the minus sign, but this comes at a cost. On non-hidpi-screen the horizontal line of the plus is completely swallowed at 100% zoom.
image
When I zoom in a lot it reappears, but does not look very nice still
image
(Firefox Windows)
This looked much nicer in the old antd version 🤔

File: frontend/stylesheets/trace_view/_tracing_view.less [hunk 10]
```diff
@@ -198,16 +198,40 @@
...
+    }
+
+    .ant-btn:first-child {
+      border-radius: 2px 0 0 2px;
+    }
+
+    .ant-btn:last-child {
+      border-radius: 0 2px 2px 0;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (philippotto) [reply frontend/stylesheets/trace_view/_tracing_view.less]: good point! I moved the css class so that it can be reused :+1:

File: frontend/javascripts/admin/task/task_search_form.js (hunk 5 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/admin/task/task_search_form.js]: I think I already pushed that yesterday.

**Comment by philippotto** [comment]
> > Yes, that was me :)
> > Which browser do you use? With Chrome and Firefox it looks like this on Linux:
> > image
> > Seems alright to me. How does it look for you?
> 
> Here it looks very good, but in the td view controls the Button to load isosurfaces is slightly lower than the others. Maybe that's not really bad, but I can't unsee it. (I'm on Linux with Firefox)
> With `Space`s:
> Screenshot from 2021-04-06 18-18-24
> 
> With the group:
> Screenshot from 2021-04-06 18-21-40
> 
> Probably the Tooltip around the last button messes the alignment slightly up.

Should be fixed now :)

File: frontend/javascripts/admin/tasktype/task_type_create_view.js (hunk 8 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/admin/tasktype/task_type_create_view.js:236]: @daniel-wer The blame shows, that you added this line.

File: frontend/javascripts/admin/tasktype/task_type_create_view.js (hunk 9 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/admin/tasktype/task_type_create_view.js:346]: The bug was already resolved

**Comment by philippotto** [comment]
> > > The icon that is used for expanding / collapsing is still broken for me (Firefox, Linux)
> > 
> > 
> > Hmm, did you test the newest version? The dev instance is probably not up to date. If there are still issues, let's have a call to talk about this.
> 
> What exactly do you mean by broken? The behavior is fine for me now. However, the icon itself looks weird. There is a fancy animation combining the bars of the plus sign to the minus sign, but this comes at a cost. On non-hidpi-screen the horizontal line of the plus is completely swallowed at 100% zoom.
> image
> When I zoom in a lot it reappears, but does not look very nice still
> image
> (Firefox Windows)
> This looked much nicer in the old antd version  In Chrome (Windows) it seems to work ok.

The actual button works for me, too. Regarding the rendering: it's not optimal, but shouldn't be blocking for this PR imo.

**Comment by MichaelBuessemeyer** [comment]
> > > The icon that is used for expanding / collapsing is still broken for me (Firefox, Linux)
> > 
> > 
> > Hmm, did you test the newest version? The dev instance is probably not up to date. If there are still issues, let's have a call to talk about this.
> 
> What exactly do you mean by broken? The behavior is fine for me now. However, the icon itself looks weird. There is a fancy animation combining the bars of the plus sign to the minus sign, but this comes at a cost. On non-hidpi-screen the horizontal line of the plus is completely swallowed at 100% zoom.
> image
> When I zoom in a lot it reappears, but does not look very nice still
> image
> (Firefox Windows)
> This looked much nicer in the old antd version thinking In Chrome (Windows) it seems to work ok.

Thx @fm3, that is exactly my problem. I can confirm that in firefox the issue is resolved by zooming and that chrome looks fine.
Maybe this is quite the firefox bug as the zoom causes it to render correctly :/

**Comment by philippotto** [comment]
> > > > The icon that is used for expanding / collapsing is still broken for me (Firefox, Linux)
> > > 
> > > 
> > > Hmm, did you test the newest version? The dev instance is probably not up to date. If there are still issues, let's have a call to talk about this.
> > 
> > 
> > What exactly do you mean by broken? The behavior is fine for me now. However, the icon itself looks weird. There is a fancy animation combining the bars of the plus sign to the minus sign, but this comes at a cost. On non-hidpi-screen the horizontal line of the plus is completely swallowed at 100% zoom.
> > image
> > When I zoom in a lot it reappears, but does not look very nice still
> > image
> > (Firefox Windows)
> > This looked much nicer in the old antd version thinking In Chrome (Windows) it seems to work ok.
> 
> Thx @fm3, that is exactly my problem. I can confirm that in firefox the issue is resolved by zooming and that chrome looks fine.
> Maybe this is quite the firefox bug as the zoom causes it to render correctly :/
> 
> > The actual button works for me, too. Regarding the rendering: it's not optimal, but shouldn't be blocking for this PR imo.
> 
> I agree to that

I just found the issue in the antd repo. The issue description contains a workaround: https://github.com/ant-design/ant-design/issues/29303#issue-804593263
Could you give that a try (simply add the CSS code to the antd overwrite less module)?

File: frontend/javascripts/admin/tasktype/task_type_create_view.js (hunk 8 above)
Comment (daniel-wer) [reply frontend/javascripts/admin/tasktype/task_type_create_view.js:236]: Initially, I added this because a change in one antd form item no longer would trigger a rerender of the whole form with the new version and part of the form actually needs to be rerendered if this item is changed. However, I solved this more elegantly later, using the `shouldUpdate` prop of form items. I just tested that this line is no longer needed and will remove it. Thank you for paying close attention :)

**Comment by MichaelBuessemeyer** [comment]
> I just found the issue in the antd repo. The issue description contains a workaround: [ant-design/ant-design#29303 (comment)](https://github.com/ant-design/ant-design/issues/29303#issue-804593263)
> Could you give that a try (simply add the CSS code to the antd overwrite less module)?

I am working on it, but it doesn't look like the correct solution :/

**Comment by fm3** [comment]
- [ ] changing user activation filter in user table leads to crash (»TypeError: this.state.activationFilter is null«)
 (I changed the selection here (unchecked both, and clicked ok)
image

**Comment by MichaelBuessemeyer** [comment]
> > I just found the issue in the antd repo. The issue description contains a workaround: [ant-design/ant-design#29303 (comment)](https://github.com/ant-design/ant-design/issues/29303#issue-804593263)
> > Could you give that a try (simply add the CSS code to the antd overwrite less module)?
> 
> I am working on it, but it doesn't look like the correct solution :/

I just pushed a fix that changes the CSS of the button so that it behaves exactly as in antd v3. Therefore the bug is not present anymore but the "fancy" spinning animation on toggle is gone. I think the animation is no necessary and the current solution perfectly centers the buttons content.

**Comment by MichaelBuessemeyer** [comment]
>     * [ ]  changing user activation filter in user table leads to crash (»TypeError: this.state.activationFilter is null«)
>       I changed the selection here (unchecked both, and clicked ok)
>       image

@fm3 Could you please check whether my newest commit fixes this bug?

The interface of the this kind of filters slightly changed. In antd version 3 an empty selection / reset meant an empty array and in the current version this is instead null. I simply catch this potential error with an ternary / if-else and set it manually to an empty array.

**Comment by fm3** [comment]
@MichaelBuessemeyer Works for me now :)

**Review by philippotto — APPROVED** [review]
:rocket:
